src/vtarget-out/inv-syn/inv_syn_cegar.cc

```diff
@@ -0,0 +1,962 @@
+
+  // you also need to merge CNF
+  status = cegar_status::NEXT_V;
+  return true;
+}
+
+// -------------------------------- EXTRACTIONS ------------------------------------------- //
+
+/// to extract result 
+void InvariantSynthesizerCegar::ExtractVerificationResult(bool autodet, bool pass, 
+  const std::string & vcd_file, const std::string & mod_inst_name ) {
+  if(check_in_bad_state()) return;
+
+  std::string res_file = vcd_file;
+  if (autodet) {
+    res_file = vcd_file_name;
+    pass = verification_pass;
+  }
+  ILA_WARN_IF(status != cegar_status::V_RES) << "CEGAR-loop: repeated verification step.";
+
+  vlg_mod_inst_name = vlg_mod_inst_name.empty() ? mod_inst_name : vlg_mod_inst_name;
+  if (vlg_mod_inst_name.empty()) {
+    ILA_ERROR << "Instance name in vcd is unknown and not specified";
+    bad_state = true;
+    return;
+  }
+
+  if(pass) {
+    ILA_INFO << "No counterexample has been found. CEGAR loop finishes.";
+    status = cegar_status::DONE;
+    return;
+  }
+
+  // we still need to create a verilog info analyzer
+  VerilogAnalyzer va(
+      implementation_incl_path, implementation_srcs_path,
+      vlg_mod_inst_name, implementation_top_module_name );
+
+  auto is_reg = [&](const std::string & n) -> bool {
+    if (! VerilogAnalyzerBase::is_reg(
+      va.check_hierarchical_name_type(n)))
+      return false;
+    if (design_smt_info)// if available we will use it
+      return design_smt_info->is_state_name(n);
+    return true; // if not available
+  };
+
+  // not passing
+  cex_extract = std::unique_ptr<CexExtractor>(
+    new CexExtractor(res_file,vlg_mod_inst_name, is_reg, not _vtg_config.CosaFullTrace));
+  // advance to synthesis stage
+  status = cegar_status::NEXT_S;
+} // extract result
+
+
+/// a helper function (only locally available)
+/// to extract vcd file name from the output
+std::string extract_vcd_name_from_cex(const std::string & fn) {
+  std::ifstream fin(fn);
+  if (! fin.is_open()) {
+    ILA_ERROR<<"Unable to read from:"<<fn;
+    return "";
+  }
+  std::string line;
+  while(std::getline(fin,line)) {
+    if (S_IN("*** TRACES ***" , line))
+      break;
+  }
+  if(fin.eof()) {
+    ILA_ERROR << "Cannot extract vcd filename, incorrect format, expecting *** TRACES ***";
+    return "";
+  }
+  std::string fname;
+  fin >> fname; fin >> fname; // the first is garbage;
+  ILA_ERROR_IF (! S_IN(".vcd", fname)) << "Expecting vcd file name, get " << fname;
+  return fname;
+} // extract_vcd_name_from_cex
+
+
+void InvariantSynthesizerCegar::CexGeneralizeRemoveStates(const std::vector<std::string> & n) {
+  cex_extract->DropStates(n);
+}
+
+void InvariantSynthesizerCegar::ExtractAbcSynthesisResultForEnhancement(InvariantInCnf& incremental_cnf, bool autodet, bool reachable, 
+    const std::string & given_smt_chc_result_txt) {
+  
+  if(check_in_bad_state()) return;
+
+  std::string res_file = given_smt_chc_result_txt;
+  if (autodet) {
+    reachable = cex_reachable;
+    res_file = synthesis_result_fn;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Variable 'res_file' is assigned a value that is never used.](https://app.codacy.com/app/Bo-Yuan-Huang/ILAng/pullRequest?prid=4555809)